Fix SOC DOS parsing for latest `pymatgen`

kavanase opened the pull request:
Since https://github.com/materialsproject/pymatgen/pull/4239, `Spin.down` entries are no longer stored in `pymatgen` DOS objects parsed from SOC calculations (as desired). There was a patch in `sumo` to delete this entry for SOC calculations if present, to make later downstream plotting easier, but this now causes a `KeyError` when used with the latest `pymatgen` due to this entry no longer being present:
```
sumo-dosplot
Traceback (most recent call last):
  File "/home/mmm0650/miniconda3/bin/sumo-dosplot", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/mmm0650/miniconda3/lib/python3.12/site-packages/sumo/cli/dosplot.py", line 646, in main
    dosplot(
  File "/home/mmm0650/miniconda3/lib/python3.12/site-packages/sumo/cli/dosplot.py", line 207, in dosplot
    dos, pdos = load_dos(
                ^^^^^^^^^
  File "/home/mmm0650/miniconda3/lib/python3.12/site-packages/sumo/electronic_structure/dos.py", line 132, in load_dos
    del dos.densities[Spin.down]
        ~~~~~~~~~~~~~^^^^^^^^^^^
KeyError: <Spin.down: -1>
```

This small update fixes the issue.

ajjackson commented on sumo/electronic_structure/dos.py:
```diff
@@ -125,7 +125,7 @@ def load_dos(
                 ).get_smeared_densities(gaussian)
 
-    if vr.parameters["LSORBIT"]:
+    if vr.parameters["LSORBIT"] and Spin.down in dos.densities:
```
Alternatively, this should work:
```
    if vr.parameters["LSORBIT"]:
        # pymatgen includes the spin down channel for SOC calculations, even
        # though there is no density here. We remove this channel so the
        # plotting is easier later on.
        dos.densities.pop(Spin.down, None)
        for site in dos.pdos:
            for orbital in dos.pdos[site]:
                dos.pdos[site][orbital](Spin.down, None)
```

I don't have a very strong opinion on which is more readable.

Either way, would be good to update the comment as well to clarify "pre-2025 pymatgen". Then perhaps one day somebody will read it, notice that our minimum version has increased, and delete the unnecessary code 😄

ajjackson replied:
For fun/education, here is an alternative that avoids the nested loops and unnecessary key lookups:

```python
from itertools import chain
from operator import methodcaller
remove_spin_down = methodcaller("pop", Spin.down, None)

all_orbitals = chain(*map(methodcaller("values"), dos.pdos.values()))

remove_spin_down(dos.densities)
for orbital in all_orbitals:
    remove_spin_down(orbital)
```

But it's quite a different style from most of Sumo so this might not be the best place to start using it!

kavanase replied:
Updated to the `pop` suggestion and with the `pymatgen<2025` comment

ajjackson commented:
Would it be feasible to produce a small test case for this, or are these vaspruns always monstrous?

kavanase commented:
I looked at doing this when I added the fix, but didn't see a test file for DOS loading/handling. I've added one now. I checked and found the `Cs2SnBr6` example in the repo is a SOC run so works for the test.

ajjackson commented:
That's wonderful, thanks Seán! Please add a line to the CHANGELOG and I'll be happy to merge.

PS I notice you are merging from a branch named "master". You may find that if we squash this PR or accept another one first then `git pull` stops working and you are stuck out of sync. It's salvageable, but generally easier to work on a branch with a new name instead...

kavanase commented:
Done.

Yeah should've made a new branch first, but went with the quick option. Will try remember that next time!

kavanase commented:
@ajjackson @utf just noting I've had questions from some other people who ran into this issue, so might be worth making a new release with this fix unless there's anything holding that back! :)

ajjackson commented:
Good idea! It should be up now: https://pypi.org/project/sumo/2.3.11/
